### PR TITLE
[WIP] Added Teammate interface to represent entities that can be added to to teams.

### PR DESCRIPTION
**Pull request by ryantheleach**
This should be considered a work in progress.

Created a Teammate interface that extends identifiable.

Anything identifiable can be added to a scoreboard whether it makes sense or not, entities, entitieLiving, creatures, enemies etc.

I've pointed this out to multiple people in the past, and it resulted in any Text being allowed to be a Score entry, however teams were still restricted to Users only.

The reasoning was that most of the options that apply to teams are currently ignored by entities, but with the announcement that entities being on a team may influence their targeting at Minecon (Shulkers to be precise) confirming that it wasn't an accident on Mojangs behalf that entities could be on teams.

It may now be worthwhile allowing plugins to set entities to teams in order to change their targeting behaviour themselves. 

Should all entities implement Teammate? Living entities? Arrows?

### CODE REVIEW

File: src/main/java/org/spongepowered/api/scoreboard/TeamBuilder.java [hunk 1]
```diff
@@ -125,14 +126,14 @@
...
+     * Sets the set of {@link Teammate}s on the {@link Teammate}.
      *
      * <p>By default, this is the empty set.</p>
      *
-     * @param users The set of {@link User}s on the {@link Team}
+     * @param teammates The set of {@link Teammate}s on the {@link Team}
      * @return This builder
      */
-    TeamBuilder users(Set<User> users);
+    TeamBuilder Teammates(Set<Teammate> teammates);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/scoreboard/TeamBuilder.java]: the method name should be lowercase `TeamBuilder teammates(Set<Teammate> teammates);`

File: src/main/java/org/spongepowered/api/entity/player/User.java [hunk 2]
```diff
@@ -27,21 +27,19 @@
...
+import org.spongepowered.api.scoreboard.Teammate;
 import org.spongepowered.api.service.permission.Subject;
-import org.spongepowered.api.util.Identifiable;
 
 /**
  * A User is the data usually associated with a Player that is persisted across server restarts.
  * This is in contrast to Player which represents the ingame entity associated with an online User.
  */
-public interface User extends DataHolder, Identifiable, ArmorEquipable, Tamer, DataSerializable, Subject, Carrier {
+public interface User extends DataHolder, Teammate, ArmorEquipable, Tamer, Subject {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/player/User.java]: Why did you remove `DataSerializable` and `Carrier`?
They aren't inherited via `Teammate`, are they?

**Comment by ryantheleach** [comment]
There is a problem with the implementation for this. Players added to the scoreboard have traditionally used names instead of UUID, as well as the fact that underneath any string can be a teammate. It's just a matter of what should be exposed + not using UUID's feels wrong as a name change would ruin their scoreboard entry.

https://bugs.mojang.com/browse/MC-50247

I'm tempted to just write a fix for this in sponge, and have scoreboards be UUID backed for players and strings, using offlinePlayer UUID's for the arbitrary strings that get added.

The problem then remains is compatibility.

File: src/main/java/org/spongepowered/api/scoreboard/Teammate.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.scoreboard;
+
+import org.spongepowered.api.util.Identifiable;
+
+/**
+ * A marker interface that represents an object that can be added to a team.
+ */
+public interface Teammate extends Identifiable {}
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/scoreboard/Teammate.java]: Shouldn't this be `TeamMate`?

**Comment by JBYoshi** [comment]
How does this work with `Text`s having score entries?

**Comment by ryantheleach** [comment]
> How does this work with Texts having score entries?

In terms of the API it doesn't interfere, it's only for mapping entities that can belong to teams. The only possible reason I can think of for adding arbitrary Texts to teams is either offline players, or if the coloring of playerNames on the scoreboard gets removed and setting the Team is the only way to color the fake player entries. Either way this seems like something that shouldn't be exposed in the Team API directly.

File: src/main/java/org/spongepowered/api/scoreboard/Teammate.java (hunk 3 above)
Comment (ryantheleach) [inline src/main/java/org/spongepowered/api/scoreboard/Teammate.java]: Following the formatting rules I would have thought it would be Teammate, searching online appears to be divided in what the correct usage of the word should be, but I can change it if you want.

**Comment by JBYoshi** [comment]
@ryantheleach Thanks for clarifying.

File: src/main/java/org/spongepowered/api/entity/player/User.java [hunk 4]
```diff
@@ -27,21 +27,19 @@
...
-import org.spongepowered.api.item.inventory.Carrier;
+import org.spongepowered.api.scoreboard.Teammate;
 import org.spongepowered.api.service.permission.Subject;
-import org.spongepowered.api.util.Identifiable;
 
 /**
  * A User is the data usually associated with a Player that is persisted across server restarts.
  * This is in contrast to Player which represents the ingame entity associated with an online User.
  */
-public interface User extends DataHolder, Identifiable, ArmorEquipable, Tamer, DataSerializable, Subject, Carrier {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/entity/player/User.java]: Why did you remove DataSerializeable and Carrier here?

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/player/User.java]: `DataHolder` extends `DataSerializable`
`ArmorEquipable` extends `Equipable`
`Equipable` extends `Identifiable`, `Carrier`

**Comment by Aaron1011** [comment]
Closed in https://github.com/SpongePowered/SpongeAPI/commit/2cd91adf9eb9399f9f2f7688531d5ba744b8b738.
